### DISCUSSION TITLE
Negative Jacobian From inverted Elements

### DISCUSSION THREAD

**Discussion by kevin-wandke** [idaholab/moose#21890] [Q&A Modules: Solid mechanics] [answered]
I am running a simulation of a pneumatically driven soft actuator using Moose's Tensor Mechanics and contact Modules. Despite my best efforts my simulation continually fails to converge and I routinely get "negative Jacobian" error from libmesh. 

A quick search of previous forum posts would indicate that this type of error occurs when an element becomes inverted. 

I have attempted using a shorter timestep, as well as using hex8 elements instead of tet4 ones. Additionally, if I view the element in question, it does not appear to deform much at all over the course of the simulation. However, the element does undergo large translations and rotations. Could this be causing the difficulties, or is it possible that the solver is making a bad iteration that is inverting an element temporarily? 

Please let me know if there is any other information I can provide that could be helpful in solving this issue!

I have included the error message I have received below
![image](https://user-images.githubusercontent.com/43739152/185499228-2a87ee36-5288-4bf6-affe-fde0ce46378b.png)

Location of failed element 554 within mesh 
![image](https://user-images.githubusercontent.com/43739152/185499546-82d18809-2929-4783-b068-d0d869be725d.png)

Element 554 on timestep 0 (right) and timestep 40 (left). It appears that while the element was rotated and translated, there was very little actual deformation. 
![image](https://user-images.githubusercontent.com/43739152/185500150-82c0c1dd-cc7d-4d1b-a58a-e91a049688d1.png)
 
Some parts of my input file that may be helpful in diagnosing the issue:
```
[Functions]
  [rampLinear]
    type = ParsedFunction
    value=1e5*t
 []
[]
[BCs]
     [Pressure]
        [stress]
        boundary = 'Inner_cavity'
        function=rampLinear
        use_automatic_differentiation = true
        use_displaced_mesh = true
        hht_alpha=-.25
        []
    []
[]
[Contact]
  [b1]
    primary = 'Body1 Body2 Body3 Body4 Body5 Body6 Body7 Body8'
    secondary = 'Body1C Body2C Body3C Body4C Body5C Body6C Body7C Body8C'
    model = frictionless
    penalty = 0.1e5
    formulation=penalty
    normalize_penalty=True
    normal_smoothing_distance = 0.1
    newmark_beta = 0.4
    newmark_gamma = 0.75
    []
   [superglue1]
    primary = 'SG1 SG2 SG3 SG4'
    secondary = 'SG1C SG2C SG3C  SG4C'
    model = glued
    penalty = 8e5
    formulation=penalty
    normalize_penalty=True
    normal_smoothing_distance = 0.1
    newmark_beta = 0.4
    newmark_gamma = 0.75
    []
[]
[Materials]
  [strain]
    type=ADComputeFiniteStrain
    decomposition_method=EigenSolution  
   []
  [elasticity_tensor]
    type = ADComputeElasticityTensor
    C_ijkl = '1.684e5 0.176e5 0.176e5 1.684e5 0.176e5 1.684e5 0.754e5 0.754e5 0.754e5'
    fill_method = symmetric9
  []
  [elasitc_stress]
    type = ADComputeFiniteStrainElasticStress
    mu=1e1
    kappa=1e7
  []
    [density1]
    type = GenericConstantMaterial
    prop_names = density
    prop_values = 1000e-9
  []
[]

[Executioner]
  type = Transient
  solve_type=PJFNK
  nl_rel_tol = 1e-4
  nl_abs_tol =5e-4
  l_max_its = 5
  l_tol = 1e-7
  nl_max_its = 20
  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-pc_type -sub_pc_type -pc_asm_overlap -ksp_gmres_restart'
  petsc_options_value = 'asm lu 20 151
  line_search='none'
  nl_forced_its=2
  automatic_scaling=true
  compute_scaling_once=false
  start_time=0.0
  end_time=1
  [TimeStepper]
    type = IterationAdaptiveDT
    growth_factor=1.41
    optimal_iterations = 8
    dt = .001
  []
  [TimeIntegrator]
  	type=ImplicitEuler
  []
  dtmin=.001
[]
 ```

**Comment by hugary1995** [comment, marked as answer]
For starters, I have two suggestions:
1. Make sure that there aren't any rigid body modes.
2. Try adding a damper, e.g. https://mooseframework.inl.gov/source/dampers/ElementJacobianDamper.html

Regarding the visualization you showed for the inverted element: an inverted element typically happens between nonlinear iterations, after the solution is updated by Newton/PJFNK, and that is not the final converged solution that is being output at `TIMESTEP_END`. If you really want to visualize the inverted element, you could try to output the exodus file at `NONLINEAR`.

**Reply by maxnezdyur** [reply]
For the visualization, I think in the Outputs block `additional_execute_on = FAILED` will output the final Exodus file with the inverted element without having to write every `NONLINEAR` iteration.

**Reply by GiudGiud** [reply]
Can you check the coordinates of that element? It definitely doesnt look deformed in the picture. Is it involved in contact?

**Reply by kevin-wandke** [reply]
My apologies for the late response. 

I am confident that there are not any rigid body modes as if that is the case I usually see solves fail in the first timestep or two, as opposed to later in the simulation. Furthermore, I did check and there are DirichletBC's constraining the motion in the X, Y and Z directions. 

Regarding the coordinates of the deformed element, I used the thresholding operation in paraview in order to identify the proper element ID , so I am confident that the proper element is displayed. 

In addition to implementing a damper, are there any other solver parameters that might prevent this type of error?

**Reply by GiudGiud** [reply]
Small time steps are usually how we deal with this.
The element IDs are offset by one in Paraview (starts at 1 we start at 0) so maybe the element was not the right one?
